### PR TITLE
feat: add FieldElement trait and use generic type

### PR DESCRIPTION
**Pull request by joyqvq**

### CODE REVIEW

**Review by vimwitch — CHANGES_REQUESTED** [review]
Really good stuff. I think the remaining changes to this PR will only involve the `Field` and `FieldElement` files. We can take all the basic arithmetic logic in the `Field` class and move it to the `FieldElement` itself. The `FieldElement` will need to know either the `Field` or just the prime to do this.

To avoid having state in the `FieldElement` we could have the basic arithmetic functions take the prime as an input like `fn add(&self, v: &Self, p: &Self)`. Then we have the `Field` class proxy the calls to the `FieldElement` classes by adding the prime as an argument (so the `Field` interface stays the same) like


Field:

```rust
fn add(&self, v1: &FieldElement, v2: &FieldElement) -> FieldElement {
  v1.add(v2, &self.p)
}
```

BigIntElement

```rust
fn add(&self, v: &Self, p: &Self) -> FieldElement {
  (self + v) % p
}
```

Then the subtraction/negation/modd/other functions would never accept or return negative numbers.

File: src/lib.rs [hunk 1]
```diff
@@ -118,3 +115,107 @@ extern "C" {
...
+    fn two() -> Self {
+        BigIntElement(BigInt::from(2))
+    }
+    fn one() -> Self {
+        BigIntElement(BigInt::from(1))
+    }
+    fn zero() -> Self {
+        BigIntElement(BigInt::from(0))
+    }
+    fn add(&self, v: &Self) -> Self {
```
[the remark this answers is not in the document]
Comment (vimwitch) [reply src/lib.rs:196]: Disregard most of this, i think the proxy approach described in the root comment is better.

File: src/lib.rs [hunk 2]
```diff
@@ -118,3 +115,107 @@ extern "C" {
...
+    fn from_i32(v: i32) -> Self;
+    fn to_u32(&self) -> u32;
+    fn from_bytes_le(v: &[u8]) -> Self;
+    fn to_bytes_le(&self) -> Vec<u8>;
+    fn to_bytes_le_sized(&self) -> [u8; 32];
+    fn modpow(&self, e: &Self, m: &Self) -> Self;
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, PartialOrd, Hash, Deserialize)]
+pub struct BigIntElement(pub BigInt);
```
[the remark this answers is not in the document]
Comment (joyqvq) [reply src/lib.rs:142]: yeah this way we can implement a trait for this type definition ( rust doesnt allow implement a trait for an external type). in the future if we want to replace with something else, we just put a wrapper around it and let the wrapper impl the trait

File: src/lib.rs [hunk 3]
```diff
@@ -118,3 +115,107 @@ extern "C" {
...
+
+impl FieldElement for BigIntElement {
+    fn to_u32(&self) -> u32 {
+        let (_, digits) = self.0.to_u32_digits();
+        if digits.is_empty() {
+            if digits.is_empty() {
+                return 0;
+            }
+            panic!("invalid bigint digits len for u32 conversion");
```
[the remark this answers is not in the document]
Comment (joyqvq) [reply src/lib.rs:151]: there was a typo, revised! as a next step i want to examine all panics and remove as much as possible

File: src/polynomial.rs [hunk 4]
```diff
@@ -455,13 +449,10 @@ impl<T: FieldElement> Polynomial<T> {
                 // bring the values into the field using simple arithmetic
```
Comment (vimwitch) [inline src/polynomial.rs:449]: These comments can be removed

File: src/lib.rs [hunk 5]
```diff
@@ -193,29 +186,28 @@ impl FieldElement for BigIntElement {
...
+    fn sub(&self, v: &Self, p: &Self) -> Self {
+        BigIntElement(&self.0 - &v.0).modd(p)
     }
-
-    fn neg(&self) -> Self {
-        BigIntElement(&self.0 * BigInt::from(-1))
+    fn modd(&self, p: &Self) -> Self {
+        assert!(p.0.sign() == Sign::Plus);
+        if self.0.sign() == Sign::Minus {
+            println!("modd: {:?}", &self.0);
```
Comment (vimwitch) [inline src/lib.rs:204]: Is this codepath ever being run? I'd be okay with a panic here if we're not using it.

**Review by vimwitch — CHANGES_REQUESTED** [review]
Looks great, just a couple small things.

I have modular multiplication working on the gpu and it's already faster than the cpu for batches >= 1024. We could add a `batch_mul` function to the number trait and start looking for places in the code to use it (in a different pr ofc). I think that's a simple starting approach we can use for testing, then we can start integrating more bespoke shaders in the system.

File: src/lib.rs (hunk 5 above)
Comment (joyqvq) [reply src/lib.rs:204]: this is used as a helper function for 
```
fn sub(&self, v: &Self, p: &Self) -> Self {
        BigIntElement(&self.0 + &v.0).modd(p)
    }
```
so it prob should be kept, removed println

**Pull request merged**